### PR TITLE
AutoInstrumentation add MongoDb

### PR DESCRIPTION
**Pull request by RassK**
### Background
- Re-added CallTarget instrumentation managed layer
- Added MongoDb CallTarget instrumentation
- Added MongoDb sample

### Later
- Seems MongoDb driver does some background queries (to find master instance etc) that should be filtered out, otherwise query trace is polluted with internal queries.

### CODE REVIEW

**Comment by pellared** [comment]
I made a quick look:
1. Can you remove `integrations_old.json`?
2. What do you think of adding the MongoDb "example" code to `samples/Samples.AspNetCoreMvc31` instead of creating yet another sample application? Thanks of it you would not need to modify `poc.sh` for testing MongoDB auto-instrumentation and we could quickly test more things at once.

**Comment by RassK** [comment]
> I made a quick look:
> 
> 1. Can you remove `integrations_old.json`?
> 2. What do you think of adding the MongoDb "example" code to `samples/Samples.AspNetCoreMvc31` instead of creating yet another sample application? Thanks of it you would not need to modify `poc.sh` for testing MongoDB auto-instrumentation and we could quickly test more things at once.

1. Done
2. Created new endpoint "api/mongo" which returns "internal log".

File: samples/Samples.AspNetCoreMvc31/Properties/launchSettings.json [hunk 1]
```diff
@@ -39,7 +39,12 @@
         "OTEL_DOTNET_TRACER_HOME": "$(ProjectDir)$(OutputPath)profiler-lib",
         "OTEL_INTEGRATIONS": "$(ProjectDir)$(OutputPath)profiler-lib\\integrations.json",
         "OTEL_VERSION": "1.0.0",
-        "OTEL_TRACE_HEADER_TAGS": "sample.correlation.identifier, Server"
+        "OTEL_SERVICE": "Samples.AspNetCoreMvc31",
+        "OTEL_TRACE_HEADER_TAGS": "sample.correlation.identifier, Server",
+
+        "OTEL_EXPORTER": "jaeger",
```
Comment (pjanotti) [inline samples/Samples.AspNetCoreMvc31/Properties/launchSettings.json:45]: Will this really work on the POC branch? I mean are we copying it under the sample now?

File: samples/Samples.AspNetCoreMvc31/Samples.AspNetCoreMvc31.csproj [hunk 2]
```diff
@@ -6,6 +6,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="MongoDB.Driver" Version="2.12.3" />
```
Comment (pjanotti) [inline samples/Samples.AspNetCoreMvc31/Samples.AspNetCoreMvc31.csproj:9]: Not needed.

File: src/OpenTelemetry.AutoInstrumentation.ClrProfiler.Managed/Configuration/IntegrationIds.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
+// ReSharper disable InconsistentNaming - Name is used for integration names
+namespace OpenTelemetry.AutoInstrumentation.ClrProfiler.Managed.Configuration
+{
+    internal enum IntegrationIds
+    {
+        HttpMessageHandler,
```
Comment (pjanotti) [inline src/OpenTelemetry.AutoInstrumentation.ClrProfiler.Managed/Configuration/IntegrationIds.cs:6]: I hadn't noticed this before but this looks like that to configure each instrumentation one needs to add it here. Ideally, this should code from the integrations.json. Anyway, not related to this PR per se.

File: src/OpenTelemetry.AutoInstrumentation.ClrProfiler.Managed.Core/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+[assembly: InternalsVisibleTo("Datadog.Trace.ClrProfiler.IntegrationTests, PublicKey=0024000004800000140100000602000000240000525341310008000001000100652bb911930d33a4f67cfd6c712abf77978de155dbdb07b01c08df4f5615af4753b4a5fac5d4d2143ecc8ea6aaab9b3a2fef845a2698139356bf1d1acdd4a4985d493e22b79edbdfc7d1b39de7221669b0b410fa2323a0adc264b953482dea6d913b2c6dcacb3525c3c156f2652df76a3341a39663c9c165ec3cab0f5a80ef72502e49bc5946d23be6c8583ba4a3149f4dbc9a453021e299e5e357846c40514f33db357de4c855dc3225ef96beefe6ee91da24010ea50810f919513756e67e81a363811f8581d5dd0ed4e3e33a725107e70c0e32dcd0e9038885d75f3c0a8c8e0000b98b5ca318827ecc7e5124af2a46c93d6d61c2043c242aafbd33eeb11b87")]
```
[the remark this answers is not in the document]
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation.ClrProfiler.Managed.Core/AssemblyInfo.cs:5]: removed references to test projects

File: samples/Samples.AspNetCoreMvc31/Properties/launchSettings.json (hunk 1 above)
Comment (RassK) [reply samples/Samples.AspNetCoreMvc31/Properties/launchSettings.json:45]: Yes, launch profiles do work if you run these directly. For testing purposes I can see that no launch profiles are used (there is a flag --no-launch-profile).

File: src/OpenTelemetry.AutoInstrumentation.ClrProfiler.Managed/Configuration/IntegrationIds.cs (hunk 3 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation.ClrProfiler.Managed/Configuration/IntegrationIds.cs:6]: There is a env var that enables / disables integrations without modifying integrations.json. The name here and in Integrations.json must match.

File: samples/Samples.AspNetCoreMvc31/Samples.AspNetCoreMvc31.csproj (hunk 2 above)
Comment (RassK) [reply samples/Samples.AspNetCoreMvc31/Samples.AspNetCoreMvc31.csproj:9]: This is needed for mongodb sample query. @pellared asked to integrate the sample into asp.net sample.

File: src/OpenTelemetry.AutoInstrumentation.ClrProfiler.Managed/Tagging/CommonTags.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+namespace OpenTelemetry.AutoInstrumentation.ClrProfiler.Managed.Tagging
```
[the remark this answers is not in the document]
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation.ClrProfiler.Managed/Tagging/CommonTags.cs:1]: I cleaned up more and removed references to other instrumentations. 
Do you mean also that we should try to remove entire Scoping logic?

[the remark this answers is not in the document]
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation.ClrProfiler.Managed/Tagging/CommonTags.cs:1]: Deleted whole ScopeManager and it's parts... bases on logic that `Activity.Current` could indicate scope.

File: samples/Samples.AspNetCoreMvc31/Samples.AspNetCoreMvc31.csproj (hunk 2 above)
Comment (pjanotti) [reply samples/Samples.AspNetCoreMvc31/Samples.AspNetCoreMvc31.csproj:9]: I saw references to MongoDb on samples/Samples.AspNetCoreMvc21/Samples.AspNetCoreMvc21.csproj this project here doesn't have any.

**Review by pjanotti — APPROVED** [review]
I still have the sensation that perhaps it could be trimmed a bit more, however, that is not essential to the PoC, and we can always trim later, with that in mind LGTM.

File: samples/Samples.AspNetCoreMvc31/Samples.AspNetCoreMvc31.csproj (hunk 2 above)
Comment (RassK) [reply samples/Samples.AspNetCoreMvc31/Samples.AspNetCoreMvc31.csproj:9]: The controllers from mvc21 project are linked to mvc31 project... so the mvc31 technically doesn't contain any logic.

**Comment by RassK** [comment]
> I still have the sensation that perhaps it could be trimmed a bit more, however, that is not essential to the PoC, and we can always trim later, with that in mind LGTM.

Tagging and configuration could be trimmed more, but I found it more useful than without.

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 6]
```diff
@@ -128,8 +128,7 @@ CorProfiler::Initialize(IUnknown* cor_profiler_info_unknown) {
   // Initialize ReJIT handler and define the Rewriter Callback
   if (is_calltarget_enabled) {
       rejit_handler = new RejitHandler(this->info_, [this](RejitHandlerModule* mod, RejitHandlerModuleMethod* method) {
-          return S_OK;
-          // return this->CallTarget_RewriterCallback(mod, method);
+           return this->CallTarget_RewriterCallback(mod, method);
```
Comment (pellared) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:131]: nit: remove one space 😄

**Review by pellared — COMMENTED** [review]
`./poc.sh` does not seem to cover MongoDB setup.

**Comment by pjanotti** [comment]
> ./poc.sh does not seem to cover MongoDB setup.

good point @pellared 

@RassK let's cover that on a follow-up PR automating the test, per my local test it should be just a matter of launching a docker MongoDB container

**Pull request merged**